jep/400/README.adoc

```diff
@@ -0,0 +1,224 @@
+
+It is common in the Kubernetes ecosystem to store system state in highly available persistent Kubernetes resources (e.g. Pods, Services, ConfigMaps, Secrets etc). This allows microservices running on Kubernetes to integrate via Kubernetes resources. As part of the Kubernetes native Jenkins initiative we hope to increasingly use Kubernetes resources for the storage of Jenkins state. E.g. to represent Environments, Pipeline Activity (Runs) and Releases we will use Kubernetes resources (https://kubernetes.io/docs/concepts/api-extension/custom-resources/[Custom Resources]). We intend to extend this to other Jenkins resources like MultiBranchProject, Pipeline, Run, Credentials etc.
+
+This will allow microservices on kubernetes to ‘plugin’ to Jenkins X in a similar way conceptually to the way current Jenkins java plugins but without the code having to be colocated inside the Jenkins master. Any microsevice with the right RBAC rules will be allowed to list, get, update, delete Jenkins resources via the kubernetes REST API, kubectl or the kubernetes language clients. E.g. so a ChatBot gateway component could watch for pipelines starting, changing, terminating and notify users of the state - without having to modify the Jenkins master.
+
+Another area of extensibility is via https://github.com/kubernetes/helm/blob/master/docs/charts_hooks.md[Helm hooks] which allows any https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/[kubernetes Job] to be triggered after an application is installed or upgraded to perform any kind of test to ensure the new version is good. E.g. we can use Helm Hooks to implement quality gates to ensure a release is good enough in terms of operational monitoring of failures, performance and SLA metrics - if a release is not good enough we can automatically roll back.
+
+=== UpdateBot
+
+UpdateBot is a CLI tool which automates updating downstream dependencies as part of a CI / CD pipeline by generating Pull Requests to eagerly push changes through repositories and give better feedback to upstream libraries/component developers if changes break the CI pipelines of downstream repositories. 
+
+E.g. as a developer of an upstream library, as you release new versions of your library the release pipeline can invoke UpdateBot to generate Pull Requests on all downstream repositories using your library. UpdateBot can then wait for those Pull Requests to trigger the downstream CI pipelines and pass/fail. If any downstream repositories CI jobs fail then your upstream library pipeline fails; giving valuable early feedback to both the upstream team making the change and downstream teams who become aware of which versions cause failures.
```

Comment:
The “waiting” part presumably needs to be configurable, since the downstream test suites could be quite slow in general. Preferably you could configure each upstream repository to specify a list of downstream repositories considered important enough to form a smoke test that should block upstream PRs; ideally the author of an individual upstream PR could even use some kind of chatops to request particular downstream repositories for integration tests, above and beyond the stock ones, depending on the nature of the PR.